### PR TITLE
Add mandatory tracker support

### PR DESCRIPTION
**Pull request by defagos**
# Description

This PR adds support for mandatory trackers. When creating an associated adapter with mandatory behavior, the tracker will never be affected by the `isTrackingEnabled` property anymore. This is especially useful for QoS / QoE trackers that, unlike analytics trackers, might be useful in all cases.

# API design considerations

I considered many API design options and associated signatures, from a Boolean parameter to signatures including a _when_ plus some enum parameter, but none felt natural to read. The current proposal is the best I could come up with from an expressiveness point of view.

# Changes made

- Add new `TrackingBehavior` with `optional` and `mandatory` options.
- Add `behavior` parameter to tracker adapter creation methods. Default is `optional` (behavior until now).
- Fix Commanders Act tracking issue found when performing tests (a bug could prevent a tracker from being enabled in some scenario).
- Make an existing playlist-related test less flaky.

# Checklist

- [x] APIs have been properly documented (if relevant).
- [x] The documentation has been updated (if relevant).
- [x] New unit tests have been written (if relevant).
- [x] The demo has been updated (if relevant).

### CODE REVIEW

File: Sources/Player/Types/TrackingBehavior.swift [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    /// Optional tracking.
+    ///
+    /// The ``TrackerAdapter`` takes into account the ``Player/isTrackingEnabled`` player setting.
+    case optional
+
+    /// Mandatory tracking.
+    ///
+    /// The ``TrackerAdapter`` ignores the ``Player/isTrackingEnabled`` player setting.
+    case mandatory
+}
```
Comment (waliid) [inline Sources/Player/Types/TrackingBehavior.swift:20]: Isn't a boolean sufficient? (e.g. `isTrackingMandatory` or `isTrackingOptional`)

Comment (defagos) [reply Sources/Player/Types/TrackingBehavior.swift:20]: This was my first iteration but it felt clunky IMHO. 

Moreover, and though I did not implement it, we could have additional behaviors like "tracking if the app is in the foreground only". This is something we discussed in the context of our analytics requirements and an enum can be easily extended with additional behaviors, unlike a Bool.

TL;DR I found it both more expressive and extensible, thus this proposal.

Comment (defagos) [reply Sources/Player/Types/TrackingBehavior.swift:20]: Let me elaborate a bit more about the _clunkiness_ I was mentioning.

Booleans are great at describing simple self-describing binary states like `enabled` or `visible`, for example. They are less so at describing richer behaviors that cannot be described in simple terms, like `immediate` for our seeking behavior or the tracking behavior we are discussing. The tracking behavior is here not a simple state that can be understood in itself, it is merely a characterization of the interaction between a tracker behavior and how the parent player has been configured. It is only in response to this configuration that a tracker exhibits a certain behavior, it cannot be understood in isolation.

If you look at my first commit you can maybe see why this is clunky when a simple `mandatory` Boolean is used / read in code afterwards. If the Boolean is used as is (the `isTrackingMandatory` is probably too verbose, `mandatory` is the concept here since a tracker's purpose is to track), then you would have to add it everywhere this context is missing, because just seeing a _mandatory_ Boolean does not tell you what is mandatory (is it the thing you are setting? Is it another mandatory behavior like a user consent for example?). Thus each call site would likely need to be adorned with `isTrackingMandatory` or `isUserConsentMandatory` to provide for the missing context.

It is similar with simple words like `enabled`. While they can be great at conveying _light bulb_ states, they do not tell you which of these states is actually controlled. Once several states can have these on / off states, you need to provide context when they are mixed, otherwise you don't know which one is which.

Comment (waliid) [reply Sources/Player/Types/TrackingBehavior.swift:20]: I see your point, thanks for the explanation.

File: Sources/Player/Tracking/PlayerItemTracker.swift [hunk 2]
```diff
@@ -82,37 +82,48 @@ public extension PlayerItemTracker {
     /// - Parameters:
     ///   - configuration: The tracker configuration.
     ///   - mapper: A closure that maps an item metadata to tracker metadata.
+    ///   - behavior: The tracking behavior.
```
Comment (waliid) [inline Sources/Player/Tracking/PlayerItemTracker.swift:85]:
```suggestion
    ///   - behavior: The tracking behavior.
    ///   - mapper: A closure that maps an item metadata to tracker metadata.
```

File: Sources/Player/Tracking/PlayerItemTracker.swift [hunk 3]
```diff
@@ -82,37 +82,48 @@ public extension PlayerItemTracker {
...
-    /// - Parameter mapper: A closure that maps an item metadata to tracker metadata.
+    /// - Parameters:
+    ///   - mapper: A closure that maps an item metadata to tracker metadata.
+    ///   - behavior: The tracking behavior.
     /// - Returns: The tracker adapter.
-    static func adapter<M>(mapper: @escaping (M) -> Metadata) -> TrackerAdapter<M> {
-        .init(trackerType: Self.self, configuration: (), mapper: mapper)
+    static func adapter<M>(mapper: @escaping (M) -> Metadata, behavior: TrackingBehavior = .optional) -> TrackerAdapter<M> {
+        .init(trackerType: Self.self, configuration: (), behavior: behavior, mapper: mapper)
     }
```
Comment (waliid) [inline Sources/Player/Tracking/PlayerItemTracker.swift:105]:
```suggestion
    static func adapter<M>(behavior: TrackingBehavior = .optional, mapper: @escaping (M) -> Metadata) -> TrackerAdapter<M> {
        .init(trackerType: Self.self, configuration: (), behavior: behavior, mapper: mapper)
    }
```

File: Sources/Player/Tracking/PlayerItemTracker.swift [hunk 4]
```diff
@@ -82,37 +82,48 @@ public extension PlayerItemTracker {
...
 }
 
 public extension PlayerItemTracker where Configuration == Void {
     /// Creates an adapter for the receiver.
     /// 
-    /// - Parameter mapper: A closure that maps an item metadata to tracker metadata.
+    /// - Parameters:
+    ///   - mapper: A closure that maps an item metadata to tracker metadata.
+    ///   - behavior: The tracking behavior.
```
Comment (waliid) [inline Sources/Player/Tracking/PlayerItemTracker.swift:101]:
```suggestion
    /// - Parameters:
    ///   - behavior: The tracking behavior.
    ///   - mapper: A closure that maps an item metadata to tracker metadata.
```

**Review by waliid — APPROVED** [review]

**Comment by waliid** [comment]
Great job, thank you for this PR!

File: Sources/Player/Tracking/PlayerItemTracker.swift (hunk 2 above)
Comment (defagos) [reply Sources/Player/Tracking/PlayerItemTracker.swift:85]: Fix parameter order

**Comment by defagos** [comment]
Thanks for the review and for noticing parameter odering inconsistencies. Should be fixed now.

**Review by waliid — APPROVED** [review]
